### PR TITLE
[SPARK-11968] [MLlib] : MatrixFactorizationModel recommendAll for GC times

### PR DESCRIPTION
**Pull request by rekhajoshm**
Fix for ALS recommend all methods for GC times

### CODE REVIEW

**Comment by mengxr** [comment]
This won't help much and it may cause issues because the buffer is not cleaned. It would be helpful if you can profile the implementation and show that the number of temporary objects are reduced.

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala [hunk 1]
```diff
@@ -275,15 +276,13 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
...
     val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
       case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
-        val m = srcIds.length
-        val n = dstIds.length
         val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
         var k = 0
         ratings.foreachActive { (i, j, r) =>
-          output(k) = (srcIds(i), (dstIds(j), r))
+          output.append((srcIds(i), (dstIds(j), r)))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: Is k even needed now?

**Comment by rekhajoshm** [comment]
Thanks @mengxr  Any alternative suggestion for improving upon objects needed on recommendAll functionality? I did multiple profiling/heap dump by running MatrixFactorizationModelSuite with IntelliJ/Visualvm. The GC %. used heap space and heap dumps/instances are non conclusive.thanks.
thanks @srowen , fixed for your comment.

**Comment by mengxr** [comment]
@rekhajoshm You need to do profiling on big datasets. If the improvement is not significant, then this is not the right fix. Essentially we are shuffling many small objects `(srcId, (dstId, rating))`. I don't think the fix would be trivial. We could probably see improvement if we switch the backend to DataFrame/Tungsten.

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala [hunk 2]
```diff
@@ -275,16 +276,13 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
...
+    val output = new ArrayBuffer[(Int, (Int, Double))]()
     val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
       case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
-        val m = srcIds.length
-        val n = dstIds.length
         val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
-        var k = 0
+        output.clear()
         ratings.foreachActive { (i, j, r) =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: We don't need `output` to hold the buffer. The following should work, though it doesn't really fix the GC problem:

~~~scala
        for (i <- 0 until m; j <- 0 until n) yield {
          (srcIds(i), dstIds(j), ratings(i, j))
        }
~~~

**Comment by rekhajoshm** [comment]
I concur @mengxr . Tried YourKit, and VisualVm Profiling.This does not fix the concern based on my runs with MovieLensALS and RecommendationExample. I do run into a set of other issues :-) If i do not get anything soon on this, will close this pull. thanks.

**Pull request closed without merging**